druid-widget-nursery-derive/src/prism.rs

```diff
@@ -0,0 +1,105 @@
+use syn::{parse_quote, spanned::Spanned, Data, DeriveInput, Fields, GenericParam, WherePredicate};
+
+pub fn expand_prism(input: DeriveInput) -> syn::Result<TokenStream> {
+    let variants = match input.data {
+        Data::Enum(e) => e.variants,
+        _ => panic!("this derive macro only works on structs with named fields"),
+    };
+
+    let enum_name = input.ident;
+    let enum_vis = input.vis;
+
+    let mut generics = input.generics;
+    let extra_predicates: Vec<WherePredicate> = generics
+        .params
+        .iter()
+        .filter_map(|ty| match ty {
+            GenericParam::Type(tparam) => {
+                let name = &tparam.ident;
+                Some(parse_quote! { #name: ::std::clone::Clone })
+            }
+            GenericParam::Lifetime(_) | GenericParam::Const(_) => None,
+        })
+        .collect();
+    generics
+        .make_where_clause()
+        .predicates
+        .extend(extra_predicates);
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    variants
+        .iter()
+        .map(|v| {
+            let variant_name = &v.ident;
+            let name = format_ident!("{}{}", enum_name, variant_name, span = v.span());
+
+            let inner_type;
+            let inner_expr;
+            let cloned_inner;
+            let variant_expr;
+
+            match &v.fields {
+                Fields::Named(_) => {
+                    return Err(syn::Error::new_spanned(
+                        &v,
+                        "variants with named fields are not supported for deriving `Prism`",
+                    ));
+                }
+                Fields::Unnamed(f) => {
+                    let fields = f.unnamed.iter();
+
+                    // By having the comma outside instead of inside the #(),
+                    // it is only added between items, not after the last one.
+                    // For `Variant()` the inner type is `()`, for `Variant(A)`
+                    // it is `(A)` (equal to just `A`), for `Variant(A, B)` it
+                    // is the tuple `(A, B)`.
+                    inner_type = quote! { (#(#fields),*) };
+
+                    let fields = (0..f.unnamed.len()).map(|n| format_ident!("_v{}", n + 1));
+                    let cloned = fields
+                        .clone()
+                        .map(|f| quote! { ::std::clone::Clone::clone(#f) });
+
+                    inner_expr = quote! { (#(#fields),*) };
+                    cloned_inner = quote! { (#(#cloned),*) };
+                    variant_expr = inner_expr.clone();
+                }
+                Fields::Unit => {
+                    inner_type = quote! { () };
+                    inner_expr = quote! { () };
+                    cloned_inner = quote! { () };
+                    variant_expr = quote! {};
+                }
+            }
+
+            Ok(quote! {
+                #[derive(Clone)]
+                #enum_vis struct #name;
```

Comment:
should we make it `mod $enum_name { pub struct $variant_name; }` instead of `struct $enum_name$variant_name`?
this is done in derive `Lens`

Reply:
I like that! I thought this kind of thing would require inherent associated types, great to hear it doesn't.

Reply:
Unfortunately this doesn't work because enum variants already exist in the value namespace (`Option::Some` in the value namespace is a function from `T` to `Option<T>`).

Because of https://github.com/rust-lang/rust/issues/76347, I only got errors after fully implementing this idea 🙄

Reply:
Oh by the way the way druid does it / what almost worked wasn't `mod $enum_name`, it was `mod $mod_name {}` + `impl #enum_name { pub const Some: $mod_name = $mod_name::Some; }`.